guide/src/async-await.md

```diff
@@ -0,0 +1,242 @@
+# Async / Await
+
+Both Python and Rust have support for async functions, but bindings for these functions are not as 
+straightforward as they are for blocking functions. Both languages have a fairly distinct model for 
+async functions and Python's needs in particular can sometimes be restrictive.
+
+[pyo3-asyncio](https://github.com/awestlake87/pyo3-asyncio) was created to provide conversions 
+between async Python and async Rust as well as manage the nitty gritty details of Python's event 
+loop.
+
+## Awaiting an Async Python Function in Rust
+
+Let's take a look at a dead simple async Python function:
+
+```python
+async def py_sleep():
+    await asyncio.sleep(1)
+```
+
+This function simply sleeps for 1 second and returns. So what does this look like to PyO3? 
+
+First, it helps to have a little background on Python's async functions. 
+Async functions in python are simply functions that return a `coroutine` object. You can read more about 
+`coroutine` objects in the [Python 3 docs](https://docs.python.org/3/library/asyncio-task.html), but
+for our purposes, we really don't need to know much about them. The key factor here is that calling
+an `async` function is _just like calling a regular function_, the only difference is that we have
+to do something special with the object that it returns.
+
+Normally in Python, that something special is the `await` keyword, but in Rust, we don't have the 
+luxury of using Python's syntax. Luckily, Rust also has an `await` keyword that does something
+similar, we just need to find a way of converting a `&PyAny` into a Rust future so we can use the
+`await` keyword on it.
+
+That's where pyo3-asyncio comes in. `pyo3_asyncio::into_future` performs this conversion for us:
```

Comment:
How about describing *a bit* deeper about what the function does for conversion?

Reply:
I was thinking the details might be a bit convoluted in the guide. What do you think about linking to the API docs for this function to provide the implementation details instead of putting them here inline?

Currently the API docs don't go over the nitty gritty details for this function, so I'll have to expand that a little bit. It'd probably read something like this:

> This function converts the `coroutine` into a Python `Task` using `run_coroutine_threadsafe`. A completion handler sends the result of this `Task` through a `futures::channel::oneshot::Sender<PyResult<PyObject>>` and the future returned by this function simply awaits the result through the `futures::channel::oneshot::Receiver<PyResult<PyObject>>`.

Reply:
> What do you think about linking to the API docs for this function to provide the implementation details instead of putting them here inline?

Sounds nice :+1:

Reply:
K, I'll probably have it point to the docs.rs version, so getting a proper link might have to wait until it's published to crates.io. That should happen before this PR is merged anyway I think.

Reply:
I added the future docs.rs links to the guide for many of the pyo3-asyncio functions. The only difference is that I noticed `-` and `_` are different between the crate name and the folder name in the docs so the actual links should be something like this: `https://docs.rs/pyo3-asyncio/latest/pyo3_asyncio/fn.into_future.html`. I discovered this by poking around with the async-std docs.rs links.